### ISSUE TITLE
feature request: support using this rule in its source format

### ISSUE THREAD

**Issue by qzmfranklin** [bazelbuild/rules_typescript#292]
What I want:

Git add this repo as a submodule in my repo, say, under `third_party/rules_typescript`.

Add a line in `WORKSPACE`:
```
local_repository(path='third_party/rules_typescript')
```

And it should just work.

But right now this is impossible. Something in this repo (or its dependencies) would break if I do that. I have been forced to use this rule via `yarn add @bazel/typescript`, which has a few undesirable consequences:

1. relying on running `yarn/npm` from the command line breaks the hermetcity of the build
2. `bazel` cannot run at all until `yarn` is finished, which prevents the use of `yarn_install/npm_install` (defined and recommended in rules_nodejs)
3. if I ever need to make a change to the source code of rules_typescript (for example, to make the ts_proto_library generate modules in commonjs format rather than amd format), I would have to patch the file under `node_modules/@bazel/typescript` *after* `yarn/npm` is run. This is just, ..., outrageous.

Given the above use cases, I'd like to request a feature to support using this bazel rule in its source format on top of as published artifact.

**Comment by alexeagle** [comment]
Agreed that vendoring all your dependencies into your repo ought to work.

First step that improves your experience is that the newest @bazel/typescript npm package is only JS files, and the intended installation is that you have an http_archive to fetch the bazel rules separately from installing the JS dependencies. That fixes the wrong ordering where you are forced to yarn/npm install before you can run Bazel.

After making that switch, could you post what error you get from vendoring the dependency? That would make it easier to start investigating how to fix this.

**Comment by qzmfranklin** [comment]
@alexeagle 

Thanks a lot for the nice reply.

Despite my silence these days, I have been following very closely on the development of the ABC (Angular + Bazel + Closure) project, which my team have been using.

I would definitely try what you've suggested as the first step. That is, upgrade to the latest @bazel/typescript npm pacakge. Though, at this moment, we have a few other dependencies such as rules_nodejs, rules_webtesting, rules_skylib, Angular 6.1.x, Material2 6.4.x, etc.. Basically all those that are required to run the ABC development platform.

The ABC + Material environment in China was really difficult to set up. We did it about 2 months ago. After which we had really great developer experience and engineering productivity. However, much as I would love to upgrade rules_typescript (to 0.21)  and Angular (to 7.0), as we stand right now, we cannot afford the cost of time to upgrade all our dependencies immediately. I will definitely try it out once we got over this round of release and provide more information on this and maybe some other threads.

We, as a company with a strong engineering culture and desire for developer productivity, would like to thank you guys for the great work put into the ABC project.

**Comment by alexeagle** [comment]
Hey @qzmfranklin would you like to follow-up about your Bazel usage? @mgechev is keeping track of our adopters and their success. Email me alexeagle at google dot com if you'd like to have a wider conversation about the problems you're solving.
